### PR TITLE
Add domain_name provider for netdev_stdlib

### PR DESCRIPTION
**Pull request by bmjen**

### CODE REVIEW

**Comment by bmjen** [comment]
tests/beaker_tests/netdev_stdlib/domain_name_provider_defaults.rb passed in 109.22 seconds
      Test Suite: tests @ 2015-10-20 17:28:02 -0700

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 109.22 seconds
      Average Test Time: 109.22 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

File: lib/puppet/provider/domain_name/nxapi.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def validate
+    # A general check to make sure the provided string resembles a valid domain name.
+    # Reference: Regular Expressions Cookbook, 2nd Edition, Section 8.15
+    fail ArgumentError, 'Invalid domain_name provided.' unless @resource[:name] =~ /^([a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,}$/
```
Comment (robert-w-gries) [inline lib/puppet/provider/domain_name/nxapi.rb]: Why doesn't your regexp validate ip addresses, such as 8.8.8.8?

And the following domain-name's were valid on my switch while your regexp rejects them:

```
n3k-51(config)# ip domain-name ~.aa
n3k-51(config)# show run all | sec domain-name
ip domain-name ~.aa

n3k-51(config)# ip domain-name test
n3k-51(config)# show run all | sec domain-name
ip domain-name test

n3k-51(config)# ip domain-name 123
n3k-51(config)# show run all | sec domain-name
ip domain-name 123

n3k-51(config)# ip domain-name ~
n3k-51(config)# show run all | sec domain-name
ip domain-name ~
```

Comment (bmjen) [inline lib/puppet/provider/domain_name/nxapi.rb]: I was just using a generic FQDN validation regex, as found here: https://www.safaribooksonline.com/library/view/regular-expressions-cookbook/9781449327453/ch08s15.html

However, if NXOS allows any string to be inputed as domain name, I can update this as well as the test to allow any string.

File: tests/beaker_tests/netdev_stdlib/domain_namelib.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+# Require UtilityLib.rb path.
+require File.expand_path('../../lib/utilitylib.rb', __FILE__)
+
+# A library to assist testing domain_name resource
+module DomainNameLib
+  # Group of Constants used in negative tests for domain_name provider.
+  ENSURE_NEGATIVE = 'unknown'
+  NAME_VALID      = 'test.xyz'
+  NAME_INVALID    = 'a.b.c'
```
Comment (robert-w-gries) [inline tests/beaker_tests/netdev_stdlib/domain_namelib.rb]: Remove the ENSURE_NEGATIVE and NAME_INVALID manifests

File: tests/beaker_tests/netdev_stdlib/domain_name_provider_defaults.rb [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    on(agent, cmd_str) do
+      UtilityLib.search_pattern_in_output(stdout, [/domain-name test\.xyz/],
+                                          true, self, logger)
+    end
+
+    logger.info("Check domain_name instance absence on agent :: #{result}")
+  end
+
+  # @step [Step] Requests manifest from the master server to the agent.
+  step 'TestStep :: Get invalid domain_name name manifest from master' do
```
Comment (robert-w-gries) [inline tests/beaker_tests/netdev_stdlib/domain_name_provider_defaults.rb]: Remove invalid testcase

File: tests/beaker_tests/netdev_stdlib/domain_name_provider_defaults.rb [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    on(agent, cmd_str) do
+      UtilityLib.search_pattern_in_output(stdout, [/domain-name test\.xyz/],
+                                          true, self, logger)
+    end
+
+    logger.info("Check domain_name instance absence on agent :: #{result}")
+  end
+
+  # @step [Step] Requests manifest from the master server to the agent.
+  step 'TestStep :: Get resource unknown manifest from master' do
```
Comment (robert-w-gries) [inline tests/beaker_tests/netdev_stdlib/domain_name_provider_defaults.rb]: Remove negative testcase

File: lib/puppet/provider/domain_name/nxapi.rb (hunk 1 above)
Comment (robert-w-gries) [inline lib/puppet/provider/domain_name/nxapi.rb]: Thanks!

**Comment by bmjen** [comment]
@robert-w-gries Updates made. Ready to merge?

tests/beaker_tests/netdev_stdlib/domain_name_provider_defaults.rb passed in 63.71 seconds
      Test Suite: tests @ 2015-10-21 10:04:47 -0700

      - Host Configuration Summary -


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 63.71 seconds
      Average Test Time: 63.71 seconds
              Attempted: 1
                 Passed: 1
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 1

**Pull request merged**